test-apps/display-test-app/android/imodeljs-test-app/app/src/main/java/com/bentley/imodeljs_test_app/DtaServiceAuthorizationClient.kt

```diff
@@ -0,0 +1,114 @@
+        val clientSecret = env.optStringNotEmpty("IMJS_OIDC_CLIENT_SECRET")
+        val scope = env.optStringNotEmpty("IMJS_OIDC_SCOPE")
+        val authority = env.optStringNotEmpty("IMJS_SERVICE_AUTHORITY") ?: "https://ims.bentley.com"
+        if (clientId == null || clientSecret == null || scope == null) {
+            throw Exception("env does not contain necessary config data.")
+        }
+        val scopes = scope.split(" ")
+        if (scopes.contains("openid") || scopes.contains("email") || scopes.contains("profile") || scopes.contains("organization")) {
+            throw Exception("DtaServiceAuthSettings: Scopes for a service cannot include 'openid email profile organization'")
+        }
+        return AuthSettings(clientId, clientSecret, scope, authority)
+    }
+
+    private suspend fun generateAccessToken() {
+        return withContext(Dispatchers.IO) {
+            val authURL = URL("${authSettings.authority}/connect/token")
+            with (authURL.openConnection() as HttpsURLConnection) {
+                useCaches = false
+                doOutput = true
+                requestMethod = "POST"
+                val encodedScope = Uri.encode(authSettings.scope)
+                val body = "grant_type=client_credentials&scope=${encodedScope}"
+                val bodyData = body.toByteArray()
+                setRequestProperty("Accept", "application/json")
+                setRequestProperty("Content-Type", "application/x-www-form-urlencoded")
+                setRequestProperty("charset", "utf-8")
+                setRequestProperty("Content-Length", "${bodyData.size}")
+                val encodedClientId = Uri.encode(authSettings.clientId)
+                val encodedClientSecret = Uri.encode(authSettings.clientSecret)
+                val encoded = "${encodedClientId}:${encodedClientSecret}"
+                setRequestProperty("Authorization", "Basic ${Base64.encodeToString(encoded.toByteArray(), Base64.NO_WRAP)}")
+                val os = DataOutputStream(outputStream)
+                os.write(bodyData)
+                os.flush()
+                if (responseCode != 200) {
+                    val message = "Wrong response code fetching token: $responseCode"
+                    throw Exception(message)
+                }
+                val response = JSONObject(inputStream.bufferedReader().use(BufferedReader::readText))
+                val responseValues = listOfNotNull(response.optStringNotEmpty("access_token"), response.optStringNotEmpty("expires_in"))
+                if (responseValues.size != 2) {
+                    throw Exception("Did not get token and expiration date in response.")
+                }
+                accessToken = "Bearer ${responseValues[0]}"
+                val expiresIn = responseValues[1].toLong()
+                expirationDate = Date(Date().time + expiresIn * 1000).toString()
+            }
+        }
+    }
+
+    private fun isValidToken(results: List<String>): Boolean {
```

Comment:
Can't this function just use `accessToken` and `expirationDate` directly rather than passing them in as a list?

Reply:
While strange, this was (I felt) the easiest way to verify that both values were non-null.

Reply:
I'll change it out, though.

Reply:
Actually, I still think it's the least contrived way that gets around the whole (member variables might be modified in another thread) paradigm that Kotlin has. I have to copy accessToken and expirationDate into locals, and doing it with a list like I do IMO is cleaner than the alternatives.

Reply:
I can do it another way by creating a let2 function that checks 2 variables instead of 1. Let me know if you want me to do that.